Enable port modification in state selinux.port_policy_present

driskell opened the pull request:
### What does this PR do?

Allow port_policy_present to modify the SELinux type for a port that already has a type, such as a port that exists in the system policy.

### What issues does this PR fix or reference?
#55687

### Previous Behavior
port_policy_present would call port_add_policy even though the port already exists under a different type, which calls semanage -a. This will fail if the port already exists in the policy. Default ports such as 8080 already exist in the system policy so previously you could not set additional types for any default port as it would always fail.

### New Behavior
When a port already exists in the policy, port_modify_policy is used which will call semanage -m and successfully modify the port type.

Because policy modifications results in port_get_policy finding two policies if sel_type is omitted and only returning the first, most likely the default policy and not the modification, there is first a check for an exact matching policy including sel_type, and then if that doesn't find one, it then checks for ANY policy regardless of sel_type, and thus performs a modification. If the second check finds no policy it then performs an addition.

### Tests written?
Yes

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

driskell committed:
fix: Enable port modification in state selinux.port_policy_present

driskell commented:
Couldn't run the tests locally on OSX as the requirements/dev_python27.txt doesn't seem to bring in pytest-salt, and when I tried to use requirements/tests.txt, it will now run but seems to abort... so will monitor this if they fail in the CI.

driskell commented:
Not sure why tests aren't running:
`ERROR: Could not find credentials entry with ID 'AWS_ACCESS_KEY_ID'`

bryceml commented:
> Not sure why tests aren't running:
> `ERROR: Could not find credentials entry with ID 'AWS_ACCESS_KEY_ID'`

This is because our CI is not configured to work against that branch.  You should change your pr to go into master instead of 2019.2. See: https://docs.saltstack.com/en/master/topics/development/contributing.html#salt-s-branch-topology

driskell commented:
Aha my bad. I’ll rebase

duckfez commented:
Any update available on this?

bradpoulton commented:
I am in need of this fix for the selinux.port_policy_present. Thanks for your hard work!

driskell commented:
Sorry I didn't get around to the rebase and then completely forgot about it!
I've just rebased now onto master and pushed again so hopefully this unblocks things from merge for @bryceml

driskell commented:
I noticed the None appearing in test failure and it's thrown up another existing issue - that if you use name to specify protocol/port it reports "None" in the output. So I have expanded test to include the protocol and port parameter version of the call, and removes the text that can report None. Didn't want to bring the parser into the state module.

Same issue exists in fcontext but I don't want to creep the scope of the PR as it's been a long time and this really needs to be merged.

driskell commented:
Managed to get `pre-commit` to work to run the blacken/lints and also got `nox` to work from the new documentation really easily! Much less stress than before - though I didn't use docs for `pre-commit` I stole the command line from Jenkins.

Anyhows...

- Rebased on top of latest master
- Pre-commit hooks all pass and blacken / lint has completed on all modified files
- Tests are now fixed and working correctly, there were some gruesome errors in there, my apologies for that

Hopefully this is now ready to go!

driskell commented:
@dwoz Can we remove the Has-Failing-Test label? All tests are now working.

Thanks!

twangboy commented:
@driskell Please address the issues in pre-commit

driskell commented:
@twangboy Sorry I do not have bandwidth to contribute further. Pre-commit was passing in October 2020. https://jenkinsci.saltstack.com/job/pr-pre-commit/job/PR-56164/5/display/redirect Could someone at SaltStack pick this up? I have no personal need for it anymore (I maintain my own internal module now.)

andreasdijkman commented:
What do we need to do to get this PR merged? I'm running into the same problem as state in the issue and this PR.
I've tried looking through these checks but the links are dead. Also some of them are marked as required but cannot be completed. I don't think macOS, FreeBSD or Windows have SELinux capabilities. Not sure about Ubuntu/Debian though.

driskell commented:
@twangboy I managed to find some time - migrated the test to pytests folder and updated. I can see there's some dodgy change in master at the moment where protocol will be output as `None` if the policy already exists and you didn't explicitly specify the protocol and allowed it to be parsed from the `name` instead. I chose not to fix that here - it should be fixed elsewhere - specifically it likely means port and protocol parsing needs to be public from the module... but then it grays responsibilities so I think needs a discussion first

driskell commented:
@twangboy Just to let you know I've tested this now on 3006.4 as an override drop in to the tree in `_modules` and `_states` and it's all working as expected. Please do let me know if you need something to get this merged. Really keen not to lose the momentum again as this is the second time it's been ready :+1:

driskell committed:
chore: Run pre-commit on selinux state file

driskell commented:
I've split to two commits so that the lint changes from pre-commit are applied first, and then the code changes, which hopefully will ease the code review. First commit should be zero change just syntax. Second commit is the fix for this PR.

I've confirmed that the resulting code is (almost) exactly the same as the previous one I've tested. It's just no longer got line wrap on the long line (`+ f'with specified sel_type "{sel_type}", protocol "{protocol}" and port "{port}".',`) previously it was split over two lines but looks like it isn't needed to be split anymore.

dwoz commented:
@driskell Are you still wanting to get this in?

dwoz committed:
Add changelog for #55687
